### ISSUE TITLE
Higher order effects

### ISSUE THREAD

**Issue by noughtmare** [xnning/EvEff#5]
Currently in `Control.Ev.Utils` the `Exception` effect is defined with only a `throw` operation. The `catch` operation is implemented as a handler that handler the throw operation as you would expect. But this means that you cannot change how exceptions are caught after the fact. 

The `fused-effects` and `polysemy` packages both take another approach. They implement effects with both a `throw` and a `catch` operation. The challenge in this approach is that the `catch` operation needs to know about its own effect context. This is discussed in detail in this blog post: https://reasonablypolymorphic.com/blog/freer-higher-order-effects/.

In EvEff you could define it as follows.
```Haskell
data Except x e ans = Except
  { _throw :: !(forall b. Op x b e ans)
  , _catch :: !(forall b. Op (Eff e b, x -> Eff e b) b e ans)
  }
```
But there is currently no way to define these `throw`, `catch` functions and handlers like `exeptEither`:
```Haskell
throw :: Except x :? e => x -> Eff e b
catch :: Except x :? e => Eff e ans -> (x -> Eff e ans) -> Eff e ans
exceptEither :: Eff (Except x :* e) ans -> Eff e (Either x ans)
```

I don't have a practical need for this, but I am interested in exploring how such higher order effects could be implemented in EvEff.

**Comment by noughtmare** [comment]
This week I went through the "Effect handlers in Scope" paper which I think is the origin of the higher-order effects in "free"-like effect systems in Haskell and I reimplemented all their examples in EvEff to see if EvEff is as expressive as their solution. I have been able to write all their examples in EvEff using their proposed explicit scope delimiters. I will probably write a blog post about this soon. The long and short of it is that you can just define your effects slightly differently. E.g. for "higher-order" exceptions I have written this:

```haskell
data Catch x e ans = Catch
  { _bcatch :: Op () (Maybe x) e ans
  , _ecatch :: Op () () e ans
  , _throwcatch :: forall a. Op x a e ans
  }

throw' :: Catch x :? e => x -> Eff e a
throw' = perform _throwcatch

instance Buildable (Catch x) where
  build f = Catch (f _bcatch) (f _ecatch) (f _throwcatch)

catch' :: forall x e ans. Catch x :? e => Eff e ans -> (x -> Eff e ans) -> Eff e ans
catch' act hdl = do
  x <- perform _bcatch ()
  case x of
    Just y -> hdl y
    Nothing -> act <* perform @(Catch x) _ecatch ()

data CatchResult x a = CatchReturn a | CatchThrow x Int deriving Functor

catchResult :: CatchResult x a -> Either x a
catchResult (CatchReturn x) = Right x
catchResult (CatchThrow x 0) = Left x
catchResult (CatchThrow _ _) = error "Mismatched catch block"

runCatch :: Eff (Catch x :* e) ans -> Eff e (Either x ans)
runCatch = fmap catchResult . handlerRet CatchReturn
  Catch
    { _bcatch = operation \_ k -> k Nothing >>= \case
      CatchReturn x -> pure (CatchReturn x)
      CatchThrow x 0 -> k (Just x)
      CatchThrow x n -> pure (CatchThrow x (n - 1))
    , _ecatch = operation \_ k -> k () <&> \case
      CatchThrow x n -> CatchThrow x (n + 1)
      x -> x
    , _throwcatch = except \x -> pure (CatchThrow x 0)
    }
```

**Comment by noughtmare** [comment]
My full code is here: https://gist.github.com/noughtmare/ed8e87d03a3d4bc966b8cb7edbfefcc3

**Comment by daanx** [comment]
Ha, very cool! As an extra experiment, can you run the same examples when importing [Mp.Eff](https://github.com/xnning/MpEff) instead of `Ev.Eff` -- if we did a good job it should just work out of the box (but with full effect handler semantics).  

(I think there is no hackage package yet for `Mp.Eff` so it may be more work to install it; maybe easiest to just put the two needed files  (`Mp/Eff.hs` and `Mp/Util.hs`) in the same directory as the examples)

**Comment by noughtmare** [comment]
@daanx I actually ran into issues with non-scoped handlers. I worked around it by composing two handlers into one larger handler, see [these lines](https://gist.github.com/noughtmare/ed8e87d03a3d4bc966b8cb7edbfefcc3#file-scope-hs-L235-L303). For that I actually needed to make some changes to eveff, so I can't directly port it over to `Mp.Eff`.

Now that makes me wonder what the advantage is of real unscoped effects.
